cpu/rpx0xx/periph/usbdev.c

```diff
@@ -0,0 +1,860 @@
+    if (hw_ep->ep.dir == USB_EP_DIR_IN) {
+        buf_filled_bit = USBCTRL_DPRAM_EP0_IN_BUFFER_CONTROL_FULL_0_Msk;
+    }
+    uint32_t pid_bit = hw_ep->next_pid ? USBCTRL_DPRAM_EP0_IN_BUFFER_CONTROL_PID_0_Msk : 0;
+    hw_ep->next_pid = !hw_ep->next_pid;
+    uint32_t reg = 
+        (len & USBCTRL_DPRAM_EP0_IN_BUFFER_CONTROL_LENGTH_0_Msk)
+        | buf_filled_bit
+        | pid_bit
+        | USBCTRL_DPRAM_EP0_IN_BUFFER_CONTROL_AVAILABLE_0_Msk;
+    return reg;
+}
+
+static void _usbdev_setup_sync(void)
+{
+    rpx0xx_usb_ep_t *hw_ep = _get_ep(0, USB_EP_DIR_OUT);
+    DEBUG("[rpx0xx usb] Call _usbdev_setup_sync(), buf=%p, len=%u\n",
+              hw_ep->data_buf, hw_ep->data_buf_size);
+    if (hw_ep->data_buf == NULL) {
+        return;
+    }
+    assert(hw_ep->data_buf_size >= USB_SETUP_PKT_LEN);
+    memcpy(hw_ep->data_buf, USB_BUFFER_SETUP_PKT_START, USB_SETUP_PKT_LEN);
+    hw_ep->data_buf = NULL;
+    uint32_t buf_ctrl_reg = _prepare_ep_buf_ctrl(hw_ep, 0ul);
+    buf_ctrl_reg |= USBCTRL_DPRAM_EP0_IN_BUFFER_CONTROL_LAST_0_Msk;
+    _ep_buf_ctrl_reg_write(0, USB_EP_DIR_OUT, buf_ctrl_reg, ~0x00UL);
+    _hw_usb_dev.dev.epcb(&hw_ep->ep, USBDEV_EVENT_TR_COMPLETE);
+}
+
+static void _usbdev_buffer_sync(usbdev_ep_t *ep)
+{
+    rpx0xx_usb_ep_t *hw_ep = _get_ep(ep->num, ep->dir);
+    DEBUG("[rpx0xx usb] Call _usbdev_buffer_sync(ep.num=%d, ep.dir=%s), buf=%p, len=%u",
+              hw_ep->ep.num, PRINT_DIR(hw_ep->ep.dir), hw_ep->data_buf, hw_ep->data_buf_size);
+    size_t available = _ep_get_available(&hw_ep->ep);
+    if (hw_ep->data_buf == NULL) {
+        DEBUG(" tr completed\n");
+        _hw_usb_dev.dev.epcb(&hw_ep->ep, USBDEV_EVENT_TR_COMPLETE);
+        return;
+    }
+    size_t n = (available < hw_ep->data_cnt) ? available : hw_ep->data_cnt;
+    uint8_t *bufptr = hw_ep->data_buf + hw_ep->data_buf_size - hw_ep->data_cnt;
+    assert(n <= hw_ep->data_cnt);
+    assert(bufptr >= hw_ep->data_buf);
+    assert(bufptr <= hw_ep->data_buf + hw_ep->data_buf_size);
+    if (hw_ep->ep.dir == USB_EP_DIR_IN) {
+        memcpy(hw_ep->dpram_buf, bufptr, n);
+    } else {
+        memcpy(bufptr, hw_ep->dpram_buf, n);
+    }
+    uint32_t buf_ctrl_reg = _prepare_ep_buf_ctrl(hw_ep, n);
+    if (hw_ep->data_cnt == hw_ep->data_buf_size) {
+        DEBUG(" first");
+        buf_ctrl_reg |= USBCTRL_DPRAM_EP0_IN_BUFFER_CONTROL_RESET_Msk;
+    }
+    hw_ep->data_cnt -= n;
+    if (hw_ep->data_cnt == 0) {
+        DEBUG(" last");
+        hw_ep->data_buf = NULL;
+        buf_ctrl_reg |= USBCTRL_DPRAM_EP0_IN_BUFFER_CONTROL_LAST_0_Msk;
+
+    }
+    DEBUG("\n");
+    _ep_buf_ctrl_reg_write(ep->num, ep->dir, buf_ctrl_reg, ~0x00UL);
+}
+
+static int _usbdev_ep_xmit(usbdev_ep_t *ep, uint8_t *buf, size_t len)
+{
+    rpx0xx_usb_ep_t *hw_ep = _get_ep(ep->num, ep->dir);
+
+    DEBUG("[rpx0xx usb] Call _usbdev_ep_xmit(ep.num=%d, ep.dir=%s, buf=%p, len=%u)\n",
+          ep->num, PRINT_DIR(ep->dir), buf, len);
+
+    assert(buf != NULL);
```

Comment:
You may also drop this one.
USBUS can call this function with a NULL buffer / 0 length in order to send an ACK to Host.